lib/promscrape/discovery/ec2/api.go

```diff
@@ -135,19 +254,79 @@ func getMetadataByPath(apiPath string) ([]byte, error) {
+	apiEndpoint := buildAPIEndpoint(a.region, a.endpoint, "sts")
+	data, err := getStsAPIResponse(a, apiEndpoint)
+	if err != nil {
+		return err
+	}
+	stsCredentials := &StsCredentialsResponse{}
+	if err := xml.Unmarshal(data, stsCredentials); err != nil {
+		return fmt.Errorf("cannot unmarshal sts api response from xml: %w", err)
+	}
+	t, err := time.Parse("2006-01-02T15:04:05Z", stsCredentials.AssumeRoleResult.Credentials.Expiration)
+	if err != nil {
+		return fmt.Errorf("cannot parse sts response expiration time: %w", err)
+	}
+	// update variables and expiration time
+	a.accessKey = stsCredentials.AssumeRoleResult.Credentials.AccessKeyId
+	a.secretKey = stsCredentials.AssumeRoleResult.Credentials.SecretAccessKey
+	a.token = stsCredentials.AssumeRoleResult.Credentials.SessionToken
+
+	// update expiration time, if it is not set or less than response from metadata api
+	if a.tokenExpiration.IsZero() || t.Unix() < a.tokenExpiration.Unix() {
+		a.tokenExpiration = t
+	}
+
+	return nil
+
+}
+
+// buildAPIEndpoint - creates endpoint for aws api access
+func buildAPIEndpoint(region, cfgEndpoint, service string) string {
 	// See https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Query-Requests.html
-	endpoint := fmt.Sprintf("https://ec2.%s.amazonaws.com/", cfg.region)
-	if len(cfg.endpoint) > 0 {
-		endpoint = cfg.endpoint
+	apiEndpoint := fmt.Sprintf("https://%s.%s.amazonaws.com/", service, region)
+	if len(cfgEndpoint) > 0 {
+		apiEndpoint = cfgEndpoint
 		// endpoint may contain only hostname. Convert it to proper url then.
-		if !strings.Contains(endpoint, "://") {
-			endpoint = "https://" + endpoint
+		if !strings.Contains(apiEndpoint, "://") {
+			apiEndpoint = "https://" + apiEndpoint
 		}
-		if !strings.HasSuffix(endpoint, "/") {
-			endpoint += "/"
+		if !strings.HasSuffix(apiEndpoint, "/") {
+			apiEndpoint += "/"
 		}
 	}
+	return apiEndpoint
+}
+
+// getStsAPIResponse makes request to aws sts api with role_arn
+// and returns temporary credentials with expiration time
+// https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html
+func getStsAPIResponse(cfg *apiConfig, endpoint string) ([]byte, error) {
+	// See https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Query-Requests.html
+	apiURL := fmt.Sprintf("%s?Action=%s", endpoint, "AssumeRole")
+	apiURL += "&Version=2011-06-15"
+	apiURL += fmt.Sprintf("&RoleArn=%s", cfg.roleARN)
+	// we have to provide unique session name for cloudtrail audit
+	apiURL += "&RoleSessionName=vmagent-ec2-discovery"
+	req, err := newSignedRequest(apiURL, "sts", cfg.region, cfg.accessKey, cfg.secretKey, cfg.token)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create signed request: %w", err)
+	}
+	resp, err := cfg.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("cannot perform http request to %q: %w", apiURL, err)
+	}
+	return readResponseBody(resp, apiURL)
+
+}
+
+// getEC2APIResponse lists ec2 instances with pagination
```

Comment:
This is misleading description for this function. The function performs EC2 API request with the given `action`.